File: 0567_permutation_in_string/check_inclusion.py
from collections import Counter


def check_inclusion(self, s1: str, s2: str) -> bool:
    """
    Time: O(s1 + s2)
    Space: O(s1 + s2)
    """
    first_count = Counter(s1)
    sec_count = Counter()
    for i, c in enumerate(s2):
        if first_count == sec_count:
            return True
        sec_count[c] += 1
        if i >= len(s1):
            sec_count[s2[i - len(s1)]] -= 1
            if sec_count[s2[i - len(s1)]] == 0:
                del sec_count[s2[i - len(s1)]]
    return first_count == sec_count

File: 0567_permutation_in_string/check_inclusion.go
package main

/**
 * Time: O(s1 + s2)
 * Space: O(s1)
 */
func checkInclusion(s1 string, s2 string) bool {
	s1Count := make(map[byte]int)
	counter := len(s1)
	for _, c := range s1 {
		s1Count[byte(c)]++
	}
	begin, end := 0, 0
	for end < len(s2) {
		val := s1Count[s2[end]]
		s1Count[s2[end]]--
		if val > 0 {
			counter--
		}
		end++
		if counter == 0 {
			return true
		}
		if end >= len(s1) {
			val := s1Count[s2[begin]]
			s1Count[s2[begin]]++
			if val >= 0 {
				counter++
			}
			begin++
		}
	}
	return counter == 0
}

func main() {

}
